### PR TITLE
Properly parse members in type declarations missing an identifier

### LINKED ISSUE #30749: Why does missing class identifier cause skipping parsing members?
**Version Used**:  master (6801607)

**Steps to Reproduce**:

Run the following test:

```c#
        [Fact]
        public void MissingNameForClassWithField()
        {
            var source = @"class
{
    int i;
}
";
            var root = SyntaxFactory.ParseSyntaxTree(source).GetRoot();

            var classDecl = root.DescendantNodes().OfType<ClassDeclarationSyntax>().Single();
            Assert.True(classDecl.Identifier.IsMissing);
            Assert.False(classDecl.OpenBraceToken.IsMissing);
            Assert.False(classDecl.CloseBraceToken.IsMissing);
        }
```

**Expected Behavior**: Test passes.

**Actual Behavior**:

Test fails on the last `Assert`. According to Roslyn Quoter, the code is parsed as:

```c#
CompilationUnit().WithMembers(List<MemberDeclarationSyntax>(new MemberDeclarationSyntax[]{
    ClassDeclaration(MissingToken(SyntaxKind.IdentifierToken))
        .WithCloseBraceToken(MissingToken(SyntaxKind.CloseBraceToken)),
    FieldDeclaration(VariableDeclaration(PredefinedType(Token(SyntaxKind.IntKeyword)))
        .WithVariables(SingletonSeparatedList<VariableDeclaratorSyntax>(
            VariableDeclarator(Identifier("i")))))
        .WithSemicolonToken(Token(
            TriviaList(), SyntaxKind.SemicolonToken,
            TriviaList(Trivia(
                SkippedTokensTrivia().WithTokens(TokenList(
                    Token(SyntaxKind.CloseBraceToken)))))))}))
```

Notice that the `class` has a missing `CloseBraceToken` and is followed by a "global" field and a skipped `CloseBraceToken`.

I think that parsing it like this is not great. Is there a reason for this?

The code that's causing this behavior is:

https://github.com/dotnet/roslyn/blob/6801607c96126ccc123c17fbd7348af30aec0b88/src/Compilers/CSharp/Portable/Parser/LanguageParser.cs#L1530-L1534

If I delete the `name.IsMissing` condition, all Roslyn tests still pass and the above test also passes.

Note that I got into a this situation while renaming a class in VS: first I deleted the old identifier and then I wanted to write the new one.

### PR DESCRIPTION
**Pull request by Rekkonnect**
Closes #30749

Note that all types with empty identifiers are merged into the same declaration.

Many and long tests were added testing out those cases and ensuring a better UX when renaming on a type declaration, or typing from scratch, or omitting an identifier by accident.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
> Additionally, this also considers every type declaration with a missing identifier as unique. Even when partial, type declarations with a missing identifier (equal to an empty string in the symbol's name) will never be merged, and they will be distinct.

What does this help with?  When would one have multiple partial members in teh same scope without a name?

**Comment by Rekkonnect** [comment]
In the event that you had multiple declarations without a name, this would cause many errors about duplicate declarations. I thought the best approach would be to consider all unnamed declarations as unique, regardless of the presence of the partial modifier.

**Comment by CyrusNajmabadi** [comment]
> In the event that you had multiple declarations without a name

When woudl this happen?  I'm strugglign to think about what scenario would cause a user to get into this state.  They'd have to go to multiple partial types in a namespace, delete the name, then move onto the next.  And so on and so forth.   When/why would that ever happen?

**Comment by Rekkonnect** [comment]
I'm not expecting this to be a common scenario either. The most likely case I can think of is doing some mass find/replace and breaking up your regex so all your class declarations end up without an identifier, totally breaking all your code, if not for bulking up the type symbols so much performance degrades.

All in all, I found this decision sound and went with it, seeing how simple changes it turned out to involve. If however you disagree, removing these changes shouldn't have a large impact in the implementation.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -1555,7 +1555,7 @@ private TypeDeclarationSyntax ParseClassOrStructOrInterfaceDeclaration(SyntaxLis
                     openBrace = this.EatToken(SyntaxKind.OpenBraceToken);
 
                     // ignore members if missing type name or missing open curly
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1557]: comment needs fixing.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 1 above)
Comment (Rekkonnect) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Fixed

**Comment by jcouv** [comment]
> }

Consider adding a test for missing identifier in presence of a primary construct parameter list as well.

---
Refers to: src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationParsingTests_MissingIdentifiers.cs:3290 in e4cce7b. [](commit_id = e4cce7b345cc00ade5c54566adeee203d0b1dfde, deletion_comment = False)

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 10) with a test suggestion

**Comment by Rekkonnect** [comment]
@jcouv for the pass over the newly added tests

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 12)

**Comment by RikkiGibson** [comment]
Thanks for the contribution @Rekkonnect!

**Pull request merged**